internal/hostname/gce/gce.go

```diff
@@ -0,0 +1,120 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package gce
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/hostname/cachedfetch"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/hostname/httputils"
+)
+
+// declare these as vars not const to ease testing
+var (
+	metadataURL = "http://169.254.169.254/computeMetadata/v1"
+)
+
+var hostnameFetcher = cachedfetch.Fetcher{
+	Name: "GCP Hostname",
+	Attempt: func(ctx context.Context) (string, error) {
+		hostname, err := getResponseWithMaxLength(ctx, metadataURL+"/instance/hostname",
+			255)
+		if err != nil {
+			return "", fmt.Errorf("unable to retrieve hostname from GCE: %s", err)
+		}
+		return hostname, nil
+	},
+}
+
+var projectIDFetcher = cachedfetch.Fetcher{
+	Name: "GCP Project ID",
+	Attempt: func(ctx context.Context) (string, error) {
+		projectID, err := getResponseWithMaxLength(ctx,
+			metadataURL+"/project/project-id",
+			255)
+		if err != nil {
+			return "", fmt.Errorf("unable to retrieve project ID from GCE: %s", err)
+		}
+		return projectID, err
+	},
+}
+
+var nameFetcher = cachedfetch.Fetcher{
+	Name: "GCP Instance Name",
+	Attempt: func(ctx context.Context) (string, error) {
+		return getResponseWithMaxLength(ctx,
+			metadataURL+"/instance/name",
+			255)
+	},
+}
+
+// GetCanonicalHostname returns the DD canonical hostname (prefer: <instance-name>.<project-id>, otherwise <hostname>)
+func GetCanonicalHostname(ctx context.Context) (string, error) {
```

Comment:
Looks like the GCE provider doesn't call `validate.ValidHostname`, as opposed to other providers. Is it on purpose? Can we refactor the validation call and try to do it only once independently of the provider? To make sure all the existing providers and the future ones get covered. (ofc the validation should allow fargate to return an empty string)

Reply:
Hmm good question. Looking at the agent code it seems like the validation logic is also skipped for GCE there as well 🤔 . I've asked the agent-shared-components team about this to try to clarify if that's intentional or just an oversight

Reply:
After talking with the agent core team this is apparently intentional since GCP doesn't follow RFC 1123. I've reached out to the container integrations team to get more details, but for now it seems this is safe